// @raydium-io/raydium-sdk/lib/cjs/base/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstructionType = exports.TxVersion = void 0;
const web3_js_1 = require("@solana/web3.js");
web3_js_1.ComputeBudgetProgram.setComputeUnitPrice;
var TxVersion;
(function (TxVersion) {
    TxVersion[TxVersion["V0"] = 0] = "V0";
    TxVersion[TxVersion["LEGACY"] = 1] = "LEGACY";
})(TxVersion = exports.TxVersion || (exports.TxVersion = {}));
var InstructionType;
(function (InstructionType) {
    InstructionType[InstructionType["createAccount"] = 0] = "createAccount";
    InstructionType[InstructionType["initAccount"] = 1] = "initAccount";
    InstructionType[InstructionType["createATA"] = 2] = "createATA";
    InstructionType[InstructionType["closeAccount"] = 3] = "closeAccount";
    InstructionType[InstructionType["transferAmount"] = 4] = "transferAmount";
    InstructionType[InstructionType["initMint"] = 5] = "initMint";
    InstructionType[InstructionType["mintTo"] = 6] = "mintTo";
    InstructionType[InstructionType["initMarket"] = 7] = "initMarket";
    InstructionType[InstructionType["util1216OwnerClaim"] = 8] = "util1216OwnerClaim";
    InstructionType[InstructionType["setComputeUnitPrice"] = 9] = "setComputeUnitPrice";
    InstructionType[InstructionType["setComputeUnitLimit"] = 10] = "setComputeUnitLimit";
    // CLMM
    InstructionType[InstructionType["clmmCreatePool"] = 11] = "clmmCreatePool";
    InstructionType[InstructionType["clmmOpenPosition"] = 12] = "clmmOpenPosition";
    InstructionType[InstructionType["clmmIncreasePosition"] = 13] = "clmmIncreasePosition";
    InstructionType[InstructionType["clmmDecreasePosition"] = 14] = "clmmDecreasePosition";
    InstructionType[InstructionType["clmmClosePosition"] = 15] = "clmmClosePosition";
    InstructionType[InstructionType["clmmSwapBaseIn"] = 16] = "clmmSwapBaseIn";
    InstructionType[InstructionType["clmmSwapBaseOut"] = 17] = "clmmSwapBaseOut";
    InstructionType[InstructionType["clmmInitReward"] = 18] = "clmmInitReward";
    InstructionType[InstructionType["clmmSetReward"] = 19] = "clmmSetReward";
    InstructionType[InstructionType["clmmCollectReward"] = 20] = "clmmCollectReward";
    InstructionType[InstructionType["ammV4Swap"] = 21] = "ammV4Swap";
    InstructionType[InstructionType["ammV4AddLiquidity"] = 22] = "ammV4AddLiquidity";
    InstructionType[InstructionType["ammV4RemoveLiquidity"] = 23] = "ammV4RemoveLiquidity";
    InstructionType[InstructionType["ammV4SimulatePoolInfo"] = 24] = "ammV4SimulatePoolInfo";
    InstructionType[InstructionType["ammV4SwapBaseIn"] = 25] = "ammV4SwapBaseIn";
    InstructionType[InstructionType["ammV4SwapBaseOut"] = 26] = "ammV4SwapBaseOut";
    InstructionType[InstructionType["ammV4CreatePool"] = 27] = "ammV4CreatePool";
    InstructionType[InstructionType["ammV4InitPool"] = 28] = "ammV4InitPool";
    InstructionType[InstructionType["ammV4CreatePoolV2"] = 29] = "ammV4CreatePoolV2";
    InstructionType[InstructionType["ammV5AddLiquidity"] = 30] = "ammV5AddLiquidity";
    InstructionType[InstructionType["ammV5RemoveLiquidity"] = 31] = "ammV5RemoveLiquidity";
    InstructionType[InstructionType["ammV5SimulatePoolInfo"] = 32] = "ammV5SimulatePoolInfo";
    InstructionType[InstructionType["ammV5SwapBaseIn"] = 33] = "ammV5SwapBaseIn";
    InstructionType[InstructionType["ammV5SwapBaseOut"] = 34] = "ammV5SwapBaseOut";
    InstructionType[InstructionType["routeSwap1"] = 35] = "routeSwap1";
    InstructionType[InstructionType["routeSwap2"] = 36] = "routeSwap2";
    InstructionType[InstructionType["farmV3Deposit"] = 37] = "farmV3Deposit";
    InstructionType[InstructionType["farmV3Withdraw"] = 38] = "farmV3Withdraw";
    InstructionType[InstructionType["farmV3CreateLedger"] = 39] = "farmV3CreateLedger";
    InstructionType[InstructionType["farmV5Deposit"] = 40] = "farmV5Deposit";
    InstructionType[InstructionType["farmV5Withdraw"] = 41] = "farmV5Withdraw";
    InstructionType[InstructionType["farmV5CreateLedger"] = 42] = "farmV5CreateLedger";
    InstructionType[InstructionType["farmV6Deposit"] = 43] = "farmV6Deposit";
    InstructionType[InstructionType["farmV6Withdraw"] = 44] = "farmV6Withdraw";
    InstructionType[InstructionType["farmV6Create"] = 45] = "farmV6Create";
    InstructionType[InstructionType["farmV6Restart"] = 46] = "farmV6Restart";
    InstructionType[InstructionType["farmV6CreatorAddReward"] = 47] = "farmV6CreatorAddReward";
    InstructionType[InstructionType["farmV6CreatorWithdraw"] = 48] = "farmV6CreatorWithdraw";
    InstructionType[InstructionType["test"] = 49] = "test";
})(InstructionType = exports.InstructionType || (exports.InstructionType = {}));
